Index: /usr/share/lipstick-jolla-home-qt5/layers/EdgeLayer.qml
===================================================================
--- .orig/usr/share/lipstick-jolla-home-qt5/layers/EdgeLayer.qml
+++ /usr/share/lipstick-jolla-home-qt5/layers/EdgeLayer.qml
@@ -30,6 +30,9 @@ Layer {
     property bool _smoothClip
     property bool _effectiveActive
 
+    property real intermediateStepThreshold: 0
+    property bool intermediateStepReached: false
+
     onActiveChanged: {
         _smoothClip = clip
         _effectiveActive = active
@@ -73,6 +76,7 @@ Layer {
             var dx = active ? x : x - _rightEdge
             var dy = active ? y : y - _bottomEdge
             if (Math.abs(dx) > threshold || Math.abs(dy) > threshold) {
+                intermediateStepReached = true
                 peekFilter.gestureTriggered()
             } else {
                 peekFilter.gestureCanceled()
@@ -250,6 +254,18 @@ Layer {
         bottomEnabled: (edgeLayer.edge == PeekFilter.Bottom && !edgeLayer.active)
                     || (edgeLayer.edge == PeekFilter.Top && edgeLayer.active)
 
+        onGestureStarted: intermediateStepReached = false
+
+        onAbsoluteProgressChanged: {
+           if (absoluteProgress == 0) {
+              return
+           } else if (edgeLayer._hideActive || (absoluteProgress > intermediateStepThreshold && !intermediateStepReached)) {
+              intermediateStepReached = true
+           } else if (absoluteProgress < intermediateStepThreshold && intermediateStepReached) {
+              intermediateStepReached = false
+           }
+        }
+
         states: [
             State {
                 name: "bottom"
Index: /usr/share/jolla-settings/pages/gestures/gestures.qml
===================================================================
--- .orig/usr/share/jolla-settings/pages/gestures/gestures.qml
+++ /usr/share/jolla-settings/pages/gestures/gestures.qml
@@ -32,6 +32,30 @@ Page {
 
             TextSwitch {
                 automaticCheck: false
+                checked: swipeToLockSettings.enabled
+                text: "Swipe to lock"
+                description: "A top edge swipe will permit to lock the screen"
+                onClicked: swipeToLockSettings.enabled = !swipeToLockSettings.enabled
+            }
+
+            Slider {
+                width: parent.width
+                enabled: swipeToLockSettings.enabled
+                handleVisible: enabled
+                label: "Lock threshold"
+                maximumValue: 90
+                minimumValue: 20
+                stepSize: 1
+                value: swipeToLockSettings.threshold
+                valueText: value + "%"
+
+                onPressAndHold: cancel()
+                onValueChanged: swipeToLockSettings.threshold = Math.round(value)
+            }
+
+
+            IconTextSwitch {
+                automaticCheck: false
                 checked: desktopSettings.left_peek_to_events
                 //% "Quick Events access"
                 text: qsTrId("settings_display-la-left_peek_at_events")
@@ -76,5 +100,13 @@ Page {
         property bool lock_screen_camera: true
     }
 
+    ConfigurationGroup {
+        id: swipeToLockSettings
+        path: "/desktop/lipstick-jolla-home/swipeToLock"
+
+        property bool enabled: true
+        property int threshold: 50
+    }
+
     DisplaySettings { id: displaySettings }
 }
Index: /usr/share/lipstick-jolla-home-qt5/layers/TopMenuLayer.qml
===================================================================
--- .orig/usr/share/lipstick-jolla-home-qt5/layers/TopMenuLayer.qml
+++ /usr/share/lipstick-jolla-home-qt5/layers/TopMenuLayer.qml
@@ -2,6 +2,7 @@ import QtQuick 2.0
 import Sailfish.Silica 1.0
 import Sailfish.Lipstick 1.0
 import org.nemomobile.lipstick 0.1
+import org.nemomobile.configuration 1.0
 
 EdgeLayer {
     id: topMenuLayer
@@ -29,6 +30,29 @@ EdgeLayer {
     hintHeight: topMenu ? topMenu.itemHeight * 2 : 0
     hintDuration: 600
 
+    property bool verticalOrientation: Lipstick.compositor.topmostWindowOrientation === Qt.PortraitOrientation
+                                       || Lipstick.compositor.topmostWindowOrientation === Qt.InvertedPortraitOrientation
+
+    ConfigurationGroup {
+        id: swipeToLockSettings
+
+        path: "/desktop/lipstick-jolla-home/swipeToLock"
+
+        property bool enabled: true
+        property int threshold: 50
+    }
+
+    intermediateStepThreshold: swipeToLockSettings.enabled ?
+                                   Math.round(
+                                       (
+                                           (verticalOrientation ?
+                                              Screen.height :
+                                              Screen.width
+                                           ) * swipeToLockSettings.threshold
+                                       ) / 100
+                                   ) :
+                                   0
+
     function show() {
         if (!active) {
             toggleActive()
Index: /usr/share/lipstick-jolla-home-qt5/topmenu/PowerButton.qml
===================================================================
--- .orig/usr/share/lipstick-jolla-home-qt5/topmenu/PowerButton.qml
+++ /usr/share/lipstick-jolla-home-qt5/topmenu/PowerButton.qml
@@ -30,7 +30,7 @@ MouseArea {
 
             anchors.centerIn: parent
 
-            source: button.down
+            source: (button.down || assumeDown)
                     ? button.iconSource + "?" + Theme.highlightColor
                     : button.iconSource
 
@@ -38,4 +38,8 @@ MouseArea {
             height: Theme.iconSizeLauncher
         }
     }
+
+    /* This is at the bottom to avoid conflicts with other patches */
+    property bool assumeDown: false
+
 }
Index: /usr/share/lipstick-jolla-home-qt5/topmenu/TopMenu.qml
===================================================================
--- .orig/usr/share/lipstick-jolla-home-qt5/topmenu/TopMenu.qml
+++ /usr/share/lipstick-jolla-home-qt5/topmenu/TopMenu.qml
@@ -30,6 +30,8 @@ SilicaFlickable {
     signal shutdown()
     signal reboot()
 
+    readonly property bool lockIndication: Lipstick.compositor.topMenuLayer.edgeFilter.active && !Lipstick.compositor.topMenuLayer.intermediateStepReached
+
     readonly property int toggleColumns: Math.floor(width / Theme.itemSizeExtraLarge)
     property int itemHeight: width/toggleColumns
     readonly property bool exposed: Lipstick.compositor.topMenuLayer.exposed
@@ -37,6 +39,19 @@ SilicaFlickable {
     property alias exposedArea: background
     property var contextMenu
 
+    /*
+     * FIXME: I'm a bit of a noob with QML, is this the right way
+     * to connect to a foreign signal?
+    */
+    Component.onCompleted: Lipstick.compositor.topMenuLayer.edgeFilter.gestureTriggered.connect(topMenu.lockScreen)
+
+    function lockScreen() {
+        if (lockIndication) {
+            Lipstick.compositor.topMenuLayer.hide()
+            Lipstick.compositor.setDisplayOff()
+        }
+    }
+
     // When menu is fully open, keep it fixed at full content height instead of clipping the
     // sections to the available screen width.
     readonly property bool expanded: !atYEnd && (Lipstick.compositor.topMenuLayer.state == "visible"
@@ -173,6 +188,8 @@ SilicaFlickable {
 
                 visible: !shutdownButton.visible || powerTransition.running
 
+                assumeDown: lockIndication
+
                 onClicked: Lipstick.compositor.setDisplayOff()
 
                 iconSource: "image://theme/graphic-display-blank"
